lib/CodeGen/CodeGenFunction.h

```diff
@@ -1238,6 +1238,48 @@ class CodeGenFunction : public CodeGenTypeCache {
   const TargetInfo &getTarget() const { return Target; }
   llvm::LLVMContext &getLLVMContext() { return CGM.getLLVMContext(); }
 
+public:
+  enum BoundsCheckKind {
+    BC_None,
+    BC_Deref,
+    BC_ArraySubscript
+  };
+
+private:
+  BoundsExpr* NextBoundsExprForCheck;
+  BoundsCheckKind NextBoundsCheckKind;
+
+public:
+  BoundsExpr* GetAndClearNextBoundsCheckExpr(BoundsCheckKind Kind) {
+    if (!getLangOpts().CheckedC)
+      return nullptr;
+
+    if (NextBoundsCheckKind == BC_None)
+      return nullptr;
+
+    if (NextBoundsCheckKind != Kind)
+      return nullptr;
```

Comment:
I was in two minds about this. It effectively depends on what expressions are between the LValueToRValue cast and the deref/array subscript expression, and what order Emit methods are called for them. That said, an assert is probably useful because it shows we are breaking a high-level assumption I have about the code.